File: testdata/golang/crypto_samples.go
// Sample for cryptobom regression testing. Mixes vulnerable, weak, strong, and
// secure-but-unrelated (crypto/rand) usage. Lives under testdata/ so the Go
// toolchain ignores it; it only needs to parse, not compile.
package samples

import (
	"crypto/des"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/md5"
	"crypto/rand"
	"crypto/rc4"
	"crypto/rsa"
	"crypto/sha1"
	"crypto/sha256"
)

func vulnerable(msg []byte) {
	// Quantum-vulnerable asymmetric crypto (key sizes / curves captured).
	rsaKey, _ := rsa.GenerateKey(rand.Reader, 2048)
	ecKey, _ := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	edPub, _, _ := ed25519.GenerateKey(rand.Reader)
	_, _ = rsa.EncryptPKCS1v15(rand.Reader, &rsaKey.PublicKey, msg)
	// Classically weak key size (flagged on top of quantum-vulnerability).
	weakRSA, _ := rsa.GenerateKey(rand.Reader, 1024)
	_, _, _ = ecKey, edPub, weakRSA
}

func weak(key, msg []byte) {
	// Weak / deprecated algorithms.
	_ = md5.Sum(msg)
	_ = sha1.New()
	_, _ = des.NewCipher(key)
	_, _ = des.NewTripleDESCipher(key)
	_, _ = rc4.NewCipher(key)
}

func strongOrInventory(msg []byte) {
	// Good usage — must NOT be flagged as a problem.
	_ = sha256.Sum256(msg)
	_, _ = rand.Read(make([]byte, 16)) // crypto/rand is the secure RNG, not a finding
}

File: testdata/csharp/CryptoSamples.cs
using System;
using System.Linq;
using System.Net.Security;
using System.Security.Authentication;
using System.Security.Cryptography;
using System.Text;

// Sample for cryptobom regression testing. .NET encodes the algorithm in the type.
namespace Example
{
    class CryptoSamples
    {
        void Vulnerable()
        {
            // Quantum-vulnerable asymmetric crypto; key size from Create(n)/ctor(n).
            var rsa = RSA.Create(2048);
            var weakRsa = new RSACryptoServiceProvider(1024);
            var ecdsa = ECDsa.Create();
            var ecdh = ECDiffieHellman.Create();
        }

        void WeakAndMisused(Aes aes)
        {
            // Weak / deprecated algorithms and an ECB misuse.
            var md5 = MD5.Create();
            var sha1 = SHA1.Create();
            var des = new DESCryptoServiceProvider();
            var tdes = TripleDES.Create();
            aes.Mode = CipherMode.ECB;
        }

        void HardcodedAndPrng(Aes aes, byte[] external)
        {
            // Hardcoded key and a variable key (the latter must NOT be flagged).
            aes.Key = Encoding.UTF8.GetBytes("hardcoded-demo-key");
            aes.IV = external;

            // Key material from System.Random — flagged via sink taint.
            var r = new Random();
            var buf = new byte[16];
            r.NextBytes(buf);
            aes.Key = buf;

            // RandomNumberGenerator is correct — must NOT be flagged.
            var good = new byte[16];
            RandomNumberGenerator.Fill(good);
            aes.IV = good;
        }

        bool TimingCompare(byte[] key, byte[] msg, byte[] provided)
        {
            var mac = new HMACSHA256(key);
            var tag = mac.ComputeHash(msg);
            // Non-constant-time comparison of a MAC — flagged.
            if (tag.SequenceEqual(provided))
            {
                return true;
            }
            // Constant-time comparison — must NOT be flagged.
            if (CryptographicOperations.FixedTimeEquals(tag, provided))
            {
                return true;
            }
            // Non-crypto comparison — must NOT be flagged.
            return msg.SequenceEqual(provided);
        }

        void TlsSetup(SslStream stream, byte[] cert)
        {
            // TLS protocol enum members: deprecated flagged, modern inventoried.
            var legacy = SslProtocols.Tls11;
            var modern = SslProtocols.Tls13;
        }

        void StrongOrInventory()
        {
            // Good usage — must NOT be flagged as a problem.
            var aes = Aes.Create();
            var sha256 = SHA256.Create();
        }
    }
}
